README.md
PrintDebug
==========

A small debug printing module that prints extra info like filenames,
function names, and line numbers. It can also recursively print objects
such as lists, tuples, and dicts.

More useful stuff may be added in the future.

Example Usage:
--------------

### Debug printing:

```python
from printdebug import debug
def myfunction():
    debug('Hello from myfunction.')

myfunction()
```

####  Output:
```
   myfile.py:    3                 myfunction(): Hello from myfunction.
```

### Formatting:

The default format for line information is `'{filename}:{lineno:>5} {name:>25}(): '`,
but can be set with the `fmt` arg:

```python
debug('Test', fmt='#{lineno} in function {name}, file: {filename}')
```

The format can be set once, and used every time with a `DebugPrinter` instance, or by overriding `printdebug.default_format`:

```python
from printdebug import DebugPrinter

dp = DebugPrinter(fmt='{filename}: {name}():#{lineno}')
dp.debug('Test')
```

There is a default colorized format, using the [`colr`](https://github.com/welbornprod/colr) module.

```python
from printdebug import DebugColrPrinter
dp = DebugColrPrinter()
dp.debug('Test')
```

#### Output:

![printdebug.DebugColrPrinter](https://welbornprod.com/dl/static/media/img/printdebug-debugcolrprinter.png)

### Print an object:

```python
from printdebug import printobject

o = {'key1': {'subkey1': 'value1', 'subkey2': 'value2'}}
printobject(o)
```

#### Output:
```
key1:
    subkey1:
        value1
    subkey2:
        value2
```

### Silencing debug prints:
`debug()` and `DebugPrinter()` can be silenced with `debug_enable(False)`:
```python
from printdebug import debug, debug_enable

debug('This will print.')

debug_enable(False)
debug('This will not.')

debug_enable()
debug('This will print now.')
```


CHANGES.txt
v0.3.5 01/21/19 -- Add debug_err() to DebugPrinter, refactored and tests added.
v0.3.4 11/25/17 -- Add `debug_err` for `DebugColrPrinter`.
v0.3.3 11/24/17 -- print/debug_object, print/debug_json, and debug_exc caller info was fixed.
v0.3.2 11/22/17 -- Add \`debug_exc\`, for traceback printing in debug.
v0.3.0 01/11/17 -- Providing get_frame and get_lineinfo, disabled/enabled properties, and more tests.
v0.2.2 11/24/16 -- Single instances can be disabled, and opt-in exceptions when debug is disabled.
v0.2.1 08/28/16 -- Continuations, alignment, and stderr by default.
v0.2.0 08/28/16 -- Added DebugColrPrinter, deprecated `back`, format changes.
v0.1.1 01/27/16 -- Small bug fixes.
v0.1.0 12/04/15 -- Format/config options, DebugPrinter() and clean up.
v0.0.5-1, 09/21/2014 -- Fixed install bug.
v0.0.5, 09/21/2014 -- Added optional file-object arg.
v0.0.4, 08/21/2014 -- Switching README format using pypandoc.
v0.0.3, 08/21/2014 -- Added printobject() ..needed some more functionality :)
v0.0.2, 08/21/2014 -- Small package fix.
v0.0.1, 08/21/2014 -- Initial release.


test/test_printdebug.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

""" test_printdebug.py
    Unit tests for printdebug.py.

    -Christopher Welborn 01-11-2017
"""

import sys
import unittest

from printdebug import (
    __version__,
    debug,
    debug_exc,
    default_format,
    DebugPrinter,
    DebugColrPrinter,
    get_frame,
    get_lineinfo,
    json_str,
    LineInfo,
    object_str,
    StdErrCatcher,
)

print('Testing PrintDebug v. {}'.format(__version__), file=sys.stderr)


class HelperTests(unittest.TestCase):

    def test_get_frame(self):
        """ get_frame should return the proper frame. """
        f = get_frame()
        self.assertEqual(
            'test_get_frame',
            f.f_code.co_name,
            msg='Failed to get correct frame.'
        )

        def nested_function():
            return get_frame()
        self.assertEqual(
            'nested_function',
            nested_function().f_code.co_name,
            msg='Failed to get correct frame for nested function.'
        )

    def test_get_lineinfo(self):
        """ get_lineinfo should return the proper function names. """
        li = get_lineinfo()
        self.assertEqual(
            'test_get_lineinfo',
            li.name,
            msg='Failed to get correct function-level info.'
        )

        def nested_function():
            return get_lineinfo()
        self.assertEqual(
            'nested_function',
            nested_function().name,
            msg='Failed to get correct nested-function-level info.'
        )


class LineInfoTests(unittest.TestCase):

    def test_LineInfo_from_frame(self):
        """ LineInfo.from_frame should create a valid LineInfo. """
        li = LineInfo.from_frame(get_frame())
        self.assertEqual(
            'test_LineInfo_from_frame',
            li.name,
            msg='Failed to get correct LineInfo from function level.'
        )

        def nested_function():
            return LineInfo.from_frame(get_frame())

        self.assertEqual(
            'nested_function',
            nested_function().name,
            msg='Failed to get correct LineInfo from function level.'
        )

    def test_LineInfo_from_level(self):
        """ LineInfo.from_level should create a valid LineInfo. """
        li = LineInfo.from_level(0)
        self.assertEqual(
            'test_LineInfo_from_level',
            li.name,
            msg='Failed to get correct LineInfo from function level.'
        )

        def nested_function():
            return LineInfo.from_level(0)

        self.assertEqual(
            'nested_function',
            nested_function().name,
            msg='Failed to get correct LineInfo from function level.'
        )


class DebugTests(unittest.TestCase):
    """ Tests for the module-level `debug` function. """

    def test_debug_funcname(self):
        """ debug outputs the correct function name. """
        with StdErrCatcher() as err:
            # Passing `file` explicitly because of green test-runners.
            # They don't play nicely with my own catchers.
            # Tell `debug` to output to whatever sys.stderr is at the time
            # of this method's compilation (a green StringIO catcher).
            debug('Test.', file=sys.stderr)
        self.assertIn(
            'test_debug_funcname',
            err.output,
            msg='Failed to output correct function name.',
        )
        self.assertIn(
            'Test.',
            err.output,
            msg='Failed to output correct text.',
        )

        def nested_function():
            debug('NestedTest.', file=sys.stderr)
        with StdErrCatcher() as nestederr:
            nested_function()
        self.assertIn(
            'nested_function',
            nestederr.output,
            msg='Failed to output correct function name for nested function.',
        )
        self.assertIn(
            'NestedTest.',
            nestederr.output,
            msg='Failed to output correct text for nested function.',
        )


class DebugPrinterTests(unittest.TestCase):
    """ Tests for the DebugPrinter class. """
    def setUp(self):
        self.dp_class = DebugPrinter
        self.class_name = self.dp_class.__name__

    def test_disable(self):
        """ DebugPrinter._enabled is set by all the properties/methods. """
        dp = self.dp_class(fmt=default_format)
        dp.disabled = False
        with StdErrCatcher() as err:
            dp.debug('Test.', file=sys.stderr)
        self.assertIn(
            'Test.',
            err.output,
            msg='Failed to enable debugprinter with disabled=False.'
        )

        dp.disabled = True
        with StdErrCatcher() as err:
            dp.debug('Test.', file=sys.stderr)
        self.assertEqual(
            '',
            err.output,
            msg='Failed to disable debugprinter with disabled=True.'
        )

        dp.enabled = True
        with StdErrCatcher() as err:
            dp.debug('Test.', file=sys.stderr)
        self.assertIn(
            'Test.',
            err.output,
            msg='Failed to enable debugprinter with enabled=True.'
        )

        dp.enabled = False
        with StdErrCatcher() as err:
            dp.debug('Test.', file=sys.stderr)
        self.assertEqual(
            '',
            err.output,
            msg='Failed to disable debugprinter with enabled=False.'
        )

        dp.enable()
        with StdErrCatcher() as err:
            dp.debug('Test.', file=sys.stderr)
        self.assertIn(
            'Test.',
            err.output,
            msg='Failed to enable debugprinter with enabled().'
        )

        dp.disable()
        with StdErrCatcher() as err:
            dp.debug('Test.', file=sys.stderr)
        self.assertEqual(
            '',
            err.output,
            msg='Failed to disable debugprinter with disable().'
        )

    def test_funcname(self):
        """ debug outputs the correct function name. """
        dp = self.dp_class(fmt=default_format)
        with StdErrCatcher() as err:
            dp.debug('Test.', file=sys.stderr)
        self.assertIn(
            'test_funcname',
            err.output,
            msg='Failed to output correct function name.',
        )
        self.assertIn(
            'Test.',
            err.output,
            msg='Failed to output correct text.',
        )

        def nested_function():
            dp = self.dp_class(fmt=default_format)
            dp.debug('NestedTest.', file=sys.stderr)

        with StdErrCatcher() as nestederr:
            nested_function()

        self.assertIn(
            'nested_function',
            nestederr.output,
            msg='Failed to output correct nested function name.',
        )
        self.assertIn(
            'NestedTest.',
            nestederr.output,
            msg='Failed to output correct text for nested function.',
        )

    def test_debug_transform(self):
        """ debug_err uses transform_err. """
        def transform(text):
            return text.upper()

        dp = self.dp_class(fmt=default_format)
        with StdErrCatcher() as err:
            dp.debug('testing this', transform=transform, file=sys.stderr)
        self.assertIn(
            'TESTING THIS',
            err.output,
            msg='Failed to use transform_err function.',
        )

    def test_debug_err_transform(self):
        """ debug_err uses transform_err. """
        def transform(text):
            return text.upper()

        dp = self.dp_class(fmt=default_format)
        with StdErrCatcher() as err:
            dp.debug_err('testing this', transform=transform, file=sys.stderr)
        self.assertIn(
            'TESTING THIS',
            err.output,
            msg='Failed to use transform_err function.',
        )


class DebugColrPrinterTests(DebugPrinterTests):
    """ Tests for the DebugColrPrinter class. """
    def setUp(self):
        self.dp_class = DebugColrPrinter
        self.class_name = self.dp_class.__name__


class PrintTests(unittest.TestCase):
    def test_json_str(self):
        """ json_str should work with valid json. """
        obj = {
            'apple': [3, 1, 2],
            'diver': ('apricot', 'banana', 'cat'),
        }
        obj['f'] = obj.copy()
        obj['f']['f'] = obj['f'].copy()
        expected = """
{
    "apple": [
        3,
        1,
        2
    ],
    "diver": [
        "apricot",
        "banana",
        "cat"
    ],
    "f": {
        "apple": [
            3,
            1,
            2
        ],
        "diver": [
            "apricot",
            "banana",
            "cat"
        ],
        "f": {
            "apple": [
                3,
                1,
                2
            ],
            "diver": [
                "apricot",
                "banana",
                "cat"
            ]
        }
    }
}
"""
        self.assertEqual(
            json_str(obj).strip(),
            expected.strip(),
            msg='json_str output failed.'
        )

    def test_object_str(self):
        """ print_object should print all builtin types. """
        obj = {
            'apple': [3, 1, 2],
            'diver': ('apricot', 'banana', 'cat'),
            'extra': {3.3, 1.1, 2.2},
            'genius': b'ascii beef dead pi',
        }
        obj['f'] = obj.copy()
        obj['f']['f'] = obj['f'].copy()
        expected = """
apple:
    1
    2
    3
diver:
    apricot
    banana
    cat
extra:
    1.1
    2.2
    3.3
f:
    apple:
        1
        2
        3
    diver:
        apricot
        banana
        cat
    extra:
        1.1
        2.2
        3.3
    f:
        apple:
            1
            2
            3
        diver:
            apricot
            banana
            cat
        extra:
            1.1
            2.2
            3.3
        genius:
            b'ascii beef dead pi'
    genius:
        b'ascii beef dead pi'
genius:
    b'ascii beef dead pi'
"""
        self.assertEqual(
            '\n'.join(object_str(obj)).strip(),
            expected.strip(),
            msg='print_object output failed.'
        )


if __name__ == '__main__':
    sys.exit(unittest.main(argv=sys.argv))


printdebug/__main__.py
#!/usr/bin/env python3
""" PrintDebug
    ...Just a small test implementation for printdebug.
    -Christopher Welborn 08-21-2014
"""
from __future__ import print_function
import sys

from .tools import (
    __version__,
    get_lineinfo,
    debug,
    debug_enable,
    debug_exc,
    debug_json,
    debug_object,
    DebugColrPrinter,
    DebugPrinter,
    printobject,
)

if sys.version_info.major < 3:
    print('Color printing is not available in Python 2.\n\n', file=sys.stderr)
    DebugColrPrinter = DebugPrinter  # noqa


def main(nameargs):
    if not nameargs:
        debug('Hello from main().')
        debug('The old functions still work.')
        return run_tests(
            align_test,
            another_func_test,
            color_format_test,
            continued_test,
            debug_err_test,
            debug_exc_test,
            debug_json_test,
            debug_object_test,
            debugprinter_test,
            disable_instance_test,
            disable_test,
            level_test,
            lineinfo_str_test,
            printobject_test,
        )
    # Run tests by name.
    test_funcs = []
    globs = globals()
    for test_name in test_names:
        for name in nameargs:
            if name in test_name:
                test_funcs.append(globs[test_name])
    if not test_funcs:
        print(
            '\nNo test functions found with: {}'.format(
                ', '.join(nameargs)
            ),
            file=sys.stderr,
        )
        return 1
    return run_tests(*test_funcs)


def align_test():
    """ debug(align=True) should align content without the lineinfo. """
    debug('Next line will be aligned with this one.')
    debug('...so is it?', align=True)
    dp = DebugPrinter()
    dp.debug('DebugPrinter, next line will be aligned.')
    dp.debug('...is it aligned?', align=True)

    try:
        cp = DebugColrPrinter()
    except ImportError as ex:
        print(ex, file=sys.stderr)
        return 0
    cp.debug('DebugColrPrinter, next line will be aligned.')
    cp.debug('...was it aligned?', align=True)
    return 0


def another_func_test():
    """ Function names are reported correctly, even when nested in another.
    """
    def a_nested_func_test():
        debug('Hello from a nested function.')
        debug('Should show from another_func_test.', level=2)
        return 0

    debug(
        'Testing format from another_func_test.',
        fmt='{name}, #{lineno} in {filename}: ')
    return 0 + a_nested_func_test()


def color_format_test():
    """ Try using the fmt argument to add colors using the colr module. """
    try:
        cp = DebugColrPrinter(basename=True)
    except ImportError as ex:
        print(
            'Unable to test color formatting, no colr module:\n{}'.format(ex)
        )
        return 1
    cp.debug('Testing DebugColrPrinter.')

    def sub_function():
        cp.debug('using the default ljustwidth.')
    sub_function()
    return 0


def continued_test():
    """ Check auto-line-continuation. """
    # Intermingling these on purpose.
    debug('Starting on stderr', end='')
    debug('Starting on stdout', end='', file=sys.stdout)
    debug(', and continuing.')
    debug(', and continuing on stdout.', file=sys.stdout)
    debug('A separate stderr line.')
    debug('A separate stdout line.', file=sys.stdout)
    dp = DebugPrinter()
    dp.debug('DebugPrinter, starting on stderr', end='')
    dp.debug('DebugPrinter, starting on stdout', end='', file=sys.stdout)
    dp.debug(', and continuing.')
    dp.debug(', and continuing.', file=sys.stdout)
    try:
        cp = DebugColrPrinter()
    except ImportError:
        return 0
    cp.debug('DebugColrPrinter, starting on stderr', end='')
    cp.debug('DebugColrrinter, starting on stdout', end='', file=sys.stdout)
    cp.debug(', and continuing.')
    cp.debug(', and continuing.', file=sys.stdout)

    return 0


def debug_err_test():
    """ Debug_err should use the correct color. """
    cp = DebugColrPrinter()
    if not hasattr(cp, 'debug_err'):
        debug('No DebugColrPrinter available.')
    cp.debug_err('This error message should be a different color.')


def debug_exc_test():
    """ Debug_exc should show the correct caller. """
    try:
        raise FileNotFoundError('just a test')
    except FileNotFoundError:
        debug_exc('From function:')

    try:
        raise FileNotFoundError('just a test')
    except FileNotFoundError:
        DebugColrPrinter().debug_exc('From debugprinter:')


def debug_json_test():
    """ Debug json should show the correct caller. """
    x = {'a': [1, 2, 3], 'b': (5, 6, 7)}
    debug_json(x)
    DebugColrPrinter().debug_json(x)


def debug_object_test():
    """ Debug object should show the correct caller. """
    x = {'a': [1, 2, 3], 'b': (5, 6, 7)}
    debug_object(x)
    DebugColrPrinter().debug_object(x)


def debugprinter_test():
    """ Debug printer uses it's config to format the line info. """
    dp = DebugPrinter(
        fmt='{filename}.{lineno}>{name}: ',
        ljustwidth=60,
        basename=False)
    dp.debug('Testing DebugPrinter with ljustwidth = 60.')
    return 0


def disable_test():
    """ debug_enable(False) should silence output from debug() and
        DebugPrinter.debug().
    """
    debug('Disabling debug.')
    debug_enable(False)
    for i in range(5):
        debug('{} If you can read this something is wrong.'.format(
            i))
    DebugPrinter().debug('DebugPrinter should not print this.')
    debug_enable()
    debug('Debug re-enabled.')
    return 0


def disable_instance_test():
    """ debugprinter.disable() should disable a single instance, while leaving
        the others untouched.
    """
    dp, dp2 = DebugColrPrinter(), DebugColrPrinter()
    dp.debug('Disabling this instance.')
    dp.disable()
    dp2.debug('The first DebugPrinter was disabled.')
    dp.debug('This should not print.')
    dp.enable()
    dp.debug('DebugPrinter re-enabled.')
    return 0


def level_test():
    """ Walk backwards through the frame levels. """
    try:
        dp = DebugColrPrinter()
    except ImportError as ex:
        # If this is because colr is not installed, ignore it.
        if ex.name != 'colr':
            raise
        dp = DebugPrinter()

    def sub_function():
        def subsub_function():
            for i in range(0, 6):
                try:
                    info = get_lineinfo(level=i)
                except ValueError:
                    break
                dp.debug(''.join((' ' * (10 - i), info.name)), level=i)
        subsub_function()
    sub_function()
    return 0


def lineinfo_str_test():
    """ get_lineinfo() returns a LineInfo, and str(LineInfo()) works. """
    l = get_lineinfo()
    print('{} --> Testing LineInfo__str__'.format(l))
    return 0


def printobject_test():
    """ printobject() correctly nests dict, iterable keys/values. """
    o = {'test': {'child': {'subchild': [1, 2, 3]}, 'child2': ('a', 1, None)}}
    print('\nTesting printobject({!r}):'.format(o))
    printobject(o, indent=4)
    return 0


def run_tests(*funcs):
    """ Print a header for each function, and then call it. """
    testlen = len(funcs)
    if testlen == len(test_names):
        testlen = 'all'
    print('\nRunning {} test {}...'.format(
        testlen,
        'function' if testlen == 1 else 'functions',
    ))
    errs = 0
    for func in funcs:
        print('\nTesting with {}:'.format(func.__name__))
        errs += func() or 0

    print('\nFinished running {} test {}. (errors: {})'.format(
        testlen,
        'function' if testlen == 1 else 'functions',
        errs,
    ))
    return errs


test_names = sorted(s for s in dir() if s.endswith('_test'))


if __name__ == '__main__':
    if ('-h' in sys.argv) or ('--help' in sys.argv):
        print("""
    Usage: printdebug [TEST_NAME]
           printdebug [-l]

    Options:
        TEST_NAME  : Text/Regex pattern for test functions to run.
        -h,--help  : Show this message and exit.
        -l,--list  : Show test function names and exit.
        -v,--version  : Show printdebug version and exit.
        """)
        sys.exit(0)
    if ('-v' in sys.argv) or ('--version' in sys.argv):
        print('printdebug v. {}'.format(__version__))
        sys.exit(0)
    if ('-l' in sys.argv) or ('--list' in sys.argv):
        print('Test functions: ({})\n    {}'.format(
            len(test_names),
            '\n    '.join(test_names)
        ))
        sys.exit(0)

    sys.exit(main(sys.argv[1:]))


printdebug/catchers.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import sys


class StdOutCatcher(object):

    """ Catches stdout for code inside the 'with' block.

        Usage:
            with StdOutCatcher(safe=True, maxlength=160) as fakestdout:
                # stdout is stored in fakestdout.output
                print('okay')
            # stdout is back to normal
            # retrieve the captured output..
            print('output was: {}'.format(fakestdout.output))
    """

    def __init__(self, safe=False, maxlength=160):
        # Use safe_output?
        self.safe = safe
        # Maximum length before trimming output
        self.maxlength = maxlength
        # Output
        self.lines = []

    def __enter__(self):
        # Replace stdout with self, stdout.write() will be self.write()
        self.oldstdout = sys.stdout
        sys.stdout = self
        return self

    def __exit__(self, exctype, value, traceback):
        # Fix stdout.
        sys.stdout = self.oldstdout
        # Allow exceptions to propogate by not returning True.
        return None

    @property
    def output(self):
        return '\n'.join(self.lines)

    @output.setter
    def output(self, value):
        self.lines = value.split('\n')

    def safe_output(self, s):
        """ Escape output and check max length, trim if needed. """

        s = repr(s)[1:-1]
        if self.maxlength and (len(s) > self.maxlength):
            s = '{} (..truncated)'.format(s[:self.maxlength])
        return s

    def write(self, s):
        s = s.strip('\n')
        if not s:
            return 0
        # Save output
        self.lines.append(self.safe_output(s) if self.safe else s)
        return len(s)


class StdErrCatcher(StdOutCatcher):

    def __enter__(self):
        # Replace stderr with self, stdout.write() will be self.write()
        self.oldstderr = sys.stderr
        sys.stderr = self
        return self

    def __exit__(self, exctype, value, traceback):
        # Fix stdout.
        sys.stderr = self.oldstderr
        # Allow exceptions to propogate by not returning True.
        return None


printdebug/__init__.py
""" PrintDebug
    ...small module that helps with debug printing.
    -Christopher Welborn 08-21-2014
"""

from .tools import (
    __version__,
    DebugColrPrinter,
    DebugNotEnabled,
    DebugPrinter,
    debug,
    debug_enable,
    debug_exc,
    debug_json,
    debug_object,
    default_colr_format,
    default_format,
    get_frame,
    get_lineinfo,
    json_str,
    object_str,
    pop_or,
    print_json,
    printobject,
    print_object,
    LineInfo,
    suppress,
)

from .catchers import (
    StdErrCatcher,
    StdOutCatcher,
)

__all__ = [
    '__version__',
    # Exported tools
    'DebugColrPrinter',
    'DebugNotEnabled',
    'DebugPrinter',
    'debug',
    'debug_enable',
    'debug_exc',
    'debug_json',
    'debug_object',
    'default_colr_format',
    'default_format',
    'get_frame',
    'get_lineinfo',
    'print_json',
    'print_object',
    'printobject',
    'LineInfo',
    # Exported extras
    'json_str',
    'object_str',
    'pop_or',
    'suppress',
    'StdErrCatcher',
    'StdOutCatcher',
]
